File: samples.vb/AdventureWorksLT.SqlServer/AdventureWorksLT/SalesOrderHeader.vb
<Computation("ComputeSalesOrderNumber")>
<Computation("ComputeTotalDue")>
<CheckConstraint("CK_SalesOrderHeader_DueDate", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_DueDate), Description:="Check constraint [DueDate] >= [OrderDate]")>
<CheckConstraint("CK_SalesOrderHeader_Freight", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_Freight), Description:="Check constraint [Freight] >= (0.00)")>
<CheckConstraint("CK_SalesOrderHeader_ShipDate", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_ShipDate), Description:="Check constraint [ShipDate] >= [OrderDate] OR [ShipDate] IS NULL")>
<CheckConstraint("CK_SalesOrderHeader_SubTotal", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_SubTotal), Description:="Check constraint [SubTotal] >= (0.00)")>
<CheckConstraint("CK_SalesOrderHeader_TaxAmt", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_TaxAmt), Description:="Check constraint [TaxAmt] >= (0.00)")>
<CheckConstraint("CK_SalesOrderHeader_Status", GetType(My.UserMessages), NameOf(My.UserMessages.CK_SalesOrderHeader_Status), Description:="Check constraint [Status] BETWEEN (1) AND (6)")>
<UniqueConstraint("AK_SalesOrderHeader_SalesOrderNumber", Description:="Unique nonclustered constraint.")>
<DbIndex("IX_SalesOrderHeader_CustomerID", Description:="Nonclustered index.")>
Public Class SalesOrderHeader
    Inherits BaseModel(Of PK)

    <DbPrimaryKey("PK_SalesOrderHeader_SalesOrderID", Description:="Clustered index created by a primary key constraint.")>
    Public NotInheritable Class PK
        Inherits CandidateKey

        Public Sub New(ByVal salesOrderID As _Int32)
            MyBase.New(salesOrderID)
        End Sub
    End Class

    Public Class Key
        Inherits Key(Of PK)

        Shared Sub New()
            Register(Function(ByVal x As Key) x.SalesOrderID, _SalesOrderID)
        End Sub

        Protected Overrides Function CreatePrimaryKey() As PK
            Return New PK(SalesOrderID)
        End Function

        Private m_SalesOrderID As _Int32
        Public Property SalesOrderID As _Int32
            Get
                Return m_SalesOrderID
            End Get
            Private Set
                m_SalesOrderID = Value
            End Set
        End Property
    End Class

    Public Shared ReadOnly _SalesOrderID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderHeader) x.SalesOrderID)
    Public Shared ReadOnly _RevisionNumber As Mounter(Of _Byte) = RegisterColumn(Function(x As SalesOrderHeader) x.RevisionNumber)
    Public Shared ReadOnly _OrderDate As Mounter(Of _DateTime) = RegisterColumn(Function(x As SalesOrderHeader) x.OrderDate)
    Public Shared ReadOnly _DueDate As Mounter(Of _DateTime) = RegisterColumn(Function(x As SalesOrderHeader) x.DueDate)
    Public Shared ReadOnly _ShipDate As Mounter(Of _DateTime) = RegisterColumn(Function(x As SalesOrderHeader) x.ShipDate)
    Public Shared ReadOnly _Status As Mounter(Of _ByteEnum(Of SalesOrderStatus)) = RegisterColumn(Function(x As SalesOrderHeader) x.Status)
    Public Shared ReadOnly _OnlineOrderFlag As Mounter(Of _Boolean) = RegisterColumn(Function(x As SalesOrderHeader) x.OnlineOrderFlag)
    Public Shared ReadOnly _SalesOrderNumber As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.SalesOrderNumber)
    Public Shared ReadOnly _PurchaseOrderNumber As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.PurchaseOrderNumber)
    Public Shared ReadOnly _AccountNumber As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.AccountNumber)
    Public Shared ReadOnly _CustomerID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderHeader) x.CustomerID)
    Public Shared ReadOnly _ShipToAddressID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderHeader) x.ShipToAddressID)
    Public Shared ReadOnly _BillToAddressID As Mounter(Of _Int32) = RegisterColumn(Function(x As SalesOrderHeader) x.BillToAddressID)
    Public Shared ReadOnly _ShipMethod As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.ShipMethod)
    Public Shared ReadOnly _CreditCardApprovalCode As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.CreditCardApprovalCode)
    Public Shared ReadOnly _SubTotal As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderHeader) x.SubTotal)
    Public Shared ReadOnly _TaxAmt As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderHeader) x.TaxAmt)
    Public Shared ReadOnly _Freight As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderHeader) x.Freight)
    Public Shared ReadOnly _TotalDue As Mounter(Of _Decimal) = RegisterColumn(Function(x As SalesOrderHeader) x.TotalDue)
    Public Shared ReadOnly _Comment As Mounter(Of _String) = RegisterColumn(Function(x As SalesOrderHeader) x.Comment)

    Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
        Return New PK(SalesOrderID)
    End Function

    Private m_FK_Customer As Customer.PK
    Public ReadOnly Property FK_Customer As Customer.PK
        Get
            If m_FK_Customer Is Nothing Then
                m_FK_Customer = New Customer.PK(CustomerID)
            End If
            Return m_FK_Customer
        End Get
    End Property

    Private m_FK_ShipToAddress As Address.PK
    Public ReadOnly Property FK_ShipToAddress As Address.PK
        Get
            If m_FK_ShipToAddress Is Nothing Then
                m_FK_ShipToAddress = New Address.PK(ShipToAddressID)
            End If
            Return m_FK_ShipToAddress
        End Get
    End Property

    Private m_FK_ShipToCustomerAddress As CustomerAddress.PK
    Public ReadOnly Property FK_ShipToCustomerAddress As CustomerAddress.PK
        Get
            If m_FK_ShipToCustomerAddress Is Nothing Then
                m_FK_ShipToCustomerAddress = New CustomerAddress.PK(CustomerID, ShipToAddressID)
            End If
            Return m_FK_ShipToCustomerAddress
        End Get
    End Property

    Private m_FK_BillToAddress As Address.PK
    Public ReadOnly Property FK_BillToAddress As Address.PK
        Get
            If m_FK_BillToAddress Is Nothing Then
                m_FK_BillToAddress = New Address.PK(BillToAddressID)
            End If
            Return m_FK_BillToAddress
        End Get
    End Property

    Private m_FK_BillToCustomerAddress As CustomerAddress.PK
    Public ReadOnly Property FK_BillToCustomerAddress As CustomerAddress.PK
        Get
            If m_FK_BillToCustomerAddress Is Nothing Then
                m_FK_BillToCustomerAddress = New CustomerAddress.PK(CustomerID, BillToAddressID)
            End If
            Return m_FK_BillToCustomerAddress
        End Get
    End Property

    Private m_SalesOrderID As _Int32
    <Identity>
    <DbColumn(Description:="Primary key.")>
    Public Property SalesOrderID As _Int32
        Get
            Return m_SalesOrderID
        End Get
        Private Set
            m_SalesOrderID = Value
        End Set
    End Property

    Private m_RevisionNumber As _Byte
    <Required>
    <DefaultValue(CByte(0), Name:="DF_SalesOrderHeader_RevisionNumber")>
    <DbColumn(Description:="Incremental number to track changes to the sales order over time.")>
    Public Property RevisionNumber As _Byte
        Get
            Return m_RevisionNumber
        End Get
        Private Set
            m_RevisionNumber = Value
        End Set
    End Property

    Private m_OrderDate As _DateTime
    <Required>
    <SqlDateTime>
    <AutoDateTime(Name:="DF_SalesOrderHeader_OrderDate")>
    <DbColumn(Description:="Dates the sales order was created.")>
    Public Property OrderDate As _DateTime
        Get
            Return m_OrderDate
        End Get
        Private Set
            m_OrderDate = Value
        End Set
    End Property

    Private m_DueDate As _DateTime
    <Required>
    <SqlDateTime>
    <DbColumn(Description:="Date the order is due to the customer.")>
    Public Property DueDate As _DateTime
        Get
            Return m_DueDate
        End Get
        Private Set
            m_DueDate = Value
        End Set
    End Property

    Private m_ShipDate As _DateTime
    <SqlDateTime>
    <DbColumn(Description:="Date the order was shipped to the customer.")>
    Public Property ShipDate As _DateTime
        Get
            Return m_ShipDate
        End Get
        Private Set
            m_ShipDate = Value
        End Set
    End Property

    Private m_Status As _ByteEnum(Of SalesOrderStatus)
    <Required>
    <DefaultValue(GetType(SalesOrderStatus), NameOf(SalesOrderStatus.InProcess), Name:="DF_SalesOrderHeader_Status")>
    <DbColumn(Description:="Order current status. 1 = In process; 2 = Approved; 3 = Backordered; 4 = Rejected; 5 = Shipped; 6 = Cancelled")>
    Public Property Status As _ByteEnum(Of SalesOrderStatus)
        Get
            Return m_Status
        End Get
        Private Set
            m_Status = Value
        End Set
    End Property

    Private m_OnlineOrderFlag As _Boolean
    <Required>
    <DefaultValue(True, Name:="DF_SalesOrderHeader_OnlineOrderFlag")>
    <DbColumn(Description:="0 = Order placed by sales person. 1 = Order placed online by customer.")>
    Public Property OnlineOrderFlag As _Boolean
        Get
            Return m_OnlineOrderFlag
        End Get
        Private Set
            m_OnlineOrderFlag = Value
        End Set
    End Property

    Private m_SalesOrderNumber As _String
    <UdtOrderNumber>
    <DbColumn(Description:="Unique sales order identification number.")>
    Public Property SalesOrderNumber As _String
        Get
            Return m_SalesOrderNumber
        End Get
        Private Set
            m_SalesOrderNumber = Value
        End Set
    End Property

    Private m_PurchaseOrderNumber As _String
    <UdtOrderNumber>
    <DbColumn(Description:="Customer purchase order number reference.")>
    Public Property PurchaseOrderNumber As _String
        Get
            Return m_PurchaseOrderNumber
        End Get
        Private Set
            m_PurchaseOrderNumber = Value
        End Set
    End Property

    Private m_AccountNumber As _String
    <UdtAccountNumber>
    <DbColumn(Description:="Financial accounting number reference.")>
    Public Property AccountNumber As _String
        Get
            Return m_AccountNumber
        End Get
        Private Set
            m_AccountNumber = Value
        End Set
    End Property

    Private m_CustomerID As _Int32
    <Required>
    <DbColumn(Description:="Customer identification number. Foreign key to Customer.CustomerID.")>
    Public Property CustomerID As _Int32
        Get
            Return m_CustomerID
        End Get
        Private Set
            m_CustomerID = Value
        End Set
    End Property

    Private m_ShipToAddressID As _Int32
    <DbColumn(Description:="The ID of the location to send goods.  Foreign key to the Address table.")>
    Public Property ShipToAddressID As _Int32
        Get
            Return m_ShipToAddressID
        End Get
        Private Set
            m_ShipToAddressID = Value
        End Set
    End Property

    Private m_BillToAddressID As _Int32
    <DbColumn(Description:="The ID of the location to send invoices.  Foreign key to the Address table.")>
    Public Property BillToAddressID As _Int32
        Get
            Return m_BillToAddressID
        End Get
        Private Set
            m_BillToAddressID = Value
        End Set
    End Property

    Private m_ShipMethod As _String
    <Required>
    <SqlNVarChar(50)>
    <DbColumn(Description:="Shipping method. Foreign key to ShipMethod.ShipMethodID.")>
    Public Property ShipMethod As _String
        Get
            Return m_ShipMethod
        End Get
        Private Set
            m_ShipMethod = Value
        End Set
    End Property

    Private m_CreditCardApprovalCode As _String
    <SqlNVarChar(15)>
    <DbColumn(Description:="Approval code provided by the credit card company.")>
    Public Property CreditCardApprovalCode As _String
        Get
            Return m_CreditCardApprovalCode
        End Get
        Private Set
            m_CreditCardApprovalCode = Value
        End Set
    End Property

    Private m_SubToal As _Decimal
    <Required>
    <SqlMoney>
    <DefaultValue(GetType(Decimal), "0", Name:="DF_SalesOrderHeader_SubTotal")>
    <DbColumn(Description:="Sales subtotal. Computed as SUM(SalesOrderDetail.LineTotal)for the appropriate SalesOrderID.")>
    Public Property SubTotal As _Decimal
        Get
            Return m_SubToal
        End Get
        Private Set
            m_SubToal = Value
        End Set
    End Property

    Private m_TaxAmt As _Decimal
    <Required>
    <SqlMoney>
    <DefaultValue(GetType(Decimal), "0", Name:="DF_SalesOrderHeader_TaxAmt")>
    <DbColumn(Description:="Tax amount.")>
    Public Property TaxAmt As _Decimal
        Get
            Return m_TaxAmt
        End Get
        Private Set
            m_TaxAmt = Value
        End Set
    End Property

    Private m_Freight As _Decimal
    <Required>
    <SqlMoney>
    <DefaultValue(GetType(Decimal), "0", Name:="DF_SalesOrderHeader_Freight")>
    <DbColumn(Description:="Shipping cost.")>
    Public Property Freight As _Decimal
        Get
            Return m_Freight
        End Get
        Private Set(value As _Decimal)
            m_Freight = value
        End Set
    End Property

    Private m_TotalDue As _Decimal
    <Required>
    <SqlMoney>
    <DbColumn(Description:="Total due from customer. Computed as Subtotal + TaxAmt + Freight.")>
    Public Property TotalDue As _Decimal
        Get
            Return m_TotalDue
        End Get
        Private Set
            m_TotalDue = Value
        End Set
    End Property

    Private m_Comment As _String
    <SqlNVarCharMax>
    <DbColumn(Description:="Sales representative comments.")>
    Public Property Comment As _String
        Get
            Return m_Comment
        End Get
        Private Set
            m_Comment = Value
        End Set
    End Property

    <_Computation>
    Private Sub ComputeSalesOrderNumber()
        SalesOrderNumber.ComputedAs((_String.[Const]("SO") + (CType(SalesOrderID, _String)).AsSqlNVarChar(23)).IfNull(_String.[Const]("*** ERROR ***")))
    End Sub

    <_Computation>
    Private Sub ComputeTotalDue()
        TotalDue.ComputedAs((SubTotal + TaxAmt + Freight).IfNull(_Decimal.[Const](0)))
    End Sub

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_DueDate As _Boolean
        Get
            Return DueDate >= OrderDate
        End Get
    End Property

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_Freight As _Boolean
        Get
            Return Freight >= _Decimal.Const(0)
        End Get
    End Property

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_ShipDate As _Boolean
        Get
            Return ShipDate >= OrderDate Or ShipDate.IsNull()
        End Get
    End Property

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_SubTotal As _Boolean
        Get
            Return SubTotal >= _Decimal.Const(0)
        End Get
    End Property

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_TaxAmt As _Boolean
        Get
            Return TaxAmt >= _Decimal.Const(0)
        End Get
    End Property

    <_CheckConstraint>
    Private ReadOnly Property CK_SalesOrderHeader_Status As _Boolean
        Get
            Return IsValid(Status)
        End Get
    End Property

    Private Shared Function IsValid(status As _ByteEnum(Of SalesOrderStatus)) As _Boolean
        Dim byteExpr = CType(status, _Byte)
        Return byteExpr >= _Byte.[Const](1) And byteExpr <= _Byte.[Const](6)
    End Function

    <_UniqueConstraint>
    Private ReadOnly Property AK_SalesOrderHeader_SalesOrderNumber As ColumnSort()
        Get
            Return New ColumnSort() {SalesOrderNumber}
        End Get
    End Property

    <_DbIndex>
    Private ReadOnly Property IX_SalesOrderHeader_CustomerID As ColumnSort()
        Get
            Return New ColumnSort() {CustomerID}
        End Get
    End Property
End Class

File: samples/AdventureWorksLT.SqlServer/AdventureWorksLT/SalesOrderHeader.cs
using DevZest.Data;
using DevZest.Data.Annotations;
using DevZest.Data.SqlServer;

namespace DevZest.Samples.AdventureWorksLT
{
    [Computation(nameof(ComputeSalesOrderNumber))]
    [Computation(nameof(ComputeTotalDue))]
    [CheckConstraint(nameof(CK_SalesOrderHeader_DueDate), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_DueDate), Description = "Check constraint [DueDate] >= [OrderDate]")]
    [CheckConstraint(nameof(CK_SalesOrderHeader_Freight), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_Freight), Description = "Check constraint [Freight] >= (0.00)")]
    [CheckConstraint(nameof(CK_SalesOrderHeader_ShipDate), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_ShipDate), Description = "Check constraint [ShipDate] >= [OrderDate] OR [ShipDate] IS NULL")]
    [CheckConstraint(nameof(CK_SalesOrderHeader_SubTotal), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_SubTotal), Description = "Check constraint [SubTotal] >= (0.00)")]
    [CheckConstraint(nameof(CK_SalesOrderHeader_TaxAmt), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_TaxAmt), Description = "Check constraint [TaxAmt] >= (0.00)")]
    [CheckConstraint(nameof(CK_SalesOrderHeader_Status), typeof(UserMessages), nameof(UserMessages.CK_SalesOrderHeader_Status), Description = "Check constraint [Status] BETWEEN (1) AND (6)")]
    [UniqueConstraint(nameof(AK_SalesOrderHeader_SalesOrderNumber), Description = "Unique nonclustered constraint.")]
    [DbIndex(nameof(IX_SalesOrderHeader_CustomerID), Description = "Nonclustered index.")]
    public class SalesOrderHeader : BaseModel<SalesOrderHeader.PK>
    {
        [DbPrimaryKey("PK_SalesOrderHeader_SalesOrderID", Description = "Clustered index created by a primary key constraint.")]
        public sealed class PK : CandidateKey
        {
            public PK(_Int32 salesOrderID)
                : base(salesOrderID)
            {
            }
        }

        public class Key : Key<PK>
        {
            static Key()
            {
                Register((Key _) => _.SalesOrderID, _SalesOrderID);
            }

            protected override PK CreatePrimaryKey()
            {
                return new PK(SalesOrderID);
            }

            public _Int32 SalesOrderID { get; private set; }
        }

        public static readonly Mounter<_Int32> _SalesOrderID = RegisterColumn((SalesOrderHeader _) => _.SalesOrderID);
        public static readonly Mounter<_Byte> _RevisionNumber = RegisterColumn((SalesOrderHeader _) => _.RevisionNumber);
        public static readonly Mounter<_DateTime> _OrderDate = RegisterColumn((SalesOrderHeader _) => _.OrderDate);
        public static readonly Mounter<_DateTime> _DueDate = RegisterColumn((SalesOrderHeader _) => _.DueDate);
        public static readonly Mounter<_DateTime> _ShipDate = RegisterColumn((SalesOrderHeader _) => _.ShipDate);
        public static readonly Mounter<_ByteEnum<SalesOrderStatus>> _Status = RegisterColumn((SalesOrderHeader _) => _.Status);
        public static readonly Mounter<_Boolean> _OnlineOrderFlag = RegisterColumn((SalesOrderHeader _) => _.OnlineOrderFlag);
        public static readonly Mounter<_String> _SalesOrderNumber = RegisterColumn((SalesOrderHeader _) => _.SalesOrderNumber);
        public static readonly Mounter<_String> _PurchaseOrderNumber = RegisterColumn((SalesOrderHeader _) => _.PurchaseOrderNumber);
        public static readonly Mounter<_String> _AccountNumber = RegisterColumn((SalesOrderHeader _) => _.AccountNumber);
        public static readonly Mounter<_Int32> _CustomerID = RegisterColumn((SalesOrderHeader _) => _.CustomerID);
        public static readonly Mounter<_Int32> _ShipToAddressID = RegisterColumn((SalesOrderHeader _) => _.ShipToAddressID);
        public static readonly Mounter<_Int32> _BillToAddressID = RegisterColumn((SalesOrderHeader _) => _.BillToAddressID);
        public static readonly Mounter<_String> _ShipMethod = RegisterColumn((SalesOrderHeader _) => _.ShipMethod);
        public static readonly Mounter<_String> _CreditCardApprovalCode = RegisterColumn((SalesOrderHeader _) => _.CreditCardApprovalCode);
        public static readonly Mounter<_Decimal> _SubTotal = RegisterColumn((SalesOrderHeader _) => _.SubTotal);
        public static readonly Mounter<_Decimal> _TaxAmt = RegisterColumn((SalesOrderHeader _) => _.TaxAmt);
        public static readonly Mounter<_Decimal> _Freight = RegisterColumn((SalesOrderHeader _) => _.Freight);
        public static readonly Mounter<_Decimal> _TotalDue = RegisterColumn((SalesOrderHeader _) => _.TotalDue);
        public static readonly Mounter<_String> _Comment = RegisterColumn((SalesOrderHeader _) => _.Comment);

        protected sealed override PK CreatePrimaryKey()
        {
            return new PK(SalesOrderID);
        }

        private Customer.PK _fk_customer;
        public Customer.PK FK_Customer
        {
            get { return _fk_customer ?? (_fk_customer = new Customer.PK(CustomerID)); }
        }

        private Address.PK _fk_shipToAddress;
        public Address.PK FK_ShipToAddress
        {
            get { return _fk_shipToAddress ?? (_fk_shipToAddress = new Address.PK(ShipToAddressID)); }
        }

        private CustomerAddress.PK _fk_shipToCustomerAddress;
        public CustomerAddress.PK FK_ShipToCustomerAddress
        {
            get { return _fk_shipToCustomerAddress ?? (_fk_shipToCustomerAddress = new CustomerAddress.PK(CustomerID, ShipToAddressID)); }
        }

        private Address.PK _fk_billToAddress;
        public Address.PK FK_BillToAddress
        {
            get { return _fk_billToAddress ?? (_fk_billToAddress = new Address.PK(BillToAddressID)); }
        }

        private CustomerAddress.PK _fk_billToCustomerAddress;
        public CustomerAddress.PK FK_BillToCustomerAddress
        {
            get { return _fk_billToCustomerAddress ?? (_fk_billToCustomerAddress = new CustomerAddress.PK(CustomerID, BillToAddressID)); }
        }

        [Identity]
        [DbColumn(Description = "Primary key.")]
        public _Int32 SalesOrderID { get; private set; }

        [Required]
        [DefaultValue((byte)0, Name = "DF_SalesOrderHeader_RevisionNumber")]
        [DbColumn(Description = "Incremental number to track changes to the sales order over time.")]
        public _Byte RevisionNumber { get; private set; }

        [Required]
        [SqlDateTime]
        [AutoDateTime(Name = "DF_SalesOrderHeader_OrderDate")]
        [DbColumn(Description = "Dates the sales order was created.")]
        public _DateTime OrderDate { get; private set; }

        [Required]
        [SqlDateTime]
        [DbColumn(Description = "Date the order is due to the customer.")]
        public _DateTime DueDate { get; private set; }

        [SqlDateTime]
        [DbColumn(Description = "Date the order was shipped to the customer.")]
        public _DateTime ShipDate { get; private set; }

        [Required]
        [DefaultValue(typeof(SalesOrderStatus), nameof(SalesOrderStatus.InProcess), Name = "DF_SalesOrderHeader_Status")]
        [DbColumn(Description = "Order current status. 1 = In process; 2 = Approved; 3 = Backordered; 4 = Rejected; 5 = Shipped; 6 = Cancelled")]
        public _ByteEnum<SalesOrderStatus> Status { get; private set; }

        [Required]
        [DefaultValue(true, Name = "DF_SalesOrderHeader_OnlineOrderFlag")]
        [DbColumn(Description = "0 = Order placed by sales person. 1 = Order placed online by customer.")]
        public _Boolean OnlineOrderFlag { get; private set; }

        [UdtOrderNumber]
        [DbColumn(Description = "Unique sales order identification number.")]
        public _String SalesOrderNumber { get; private set; }

        [UdtOrderNumber]
        [DbColumn(Description = "Customer purchase order number reference.")]
        public _String PurchaseOrderNumber { get; private set; }

        [UdtAccountNumber]
        [DbColumn(Description = "Financial accounting number reference.")]
        public _String AccountNumber { get; private set; }

        [Required]
        [DbColumn(Description = "Customer identification number. Foreign key to Customer.CustomerID.")]
        public _Int32 CustomerID { get; private set; }

        [DbColumn(Description = "The ID of the location to send goods.  Foreign key to the Address table.")]
        public _Int32 ShipToAddressID { get; private set; }

        [DbColumn(Description = "The ID of the location to send invoices.  Foreign key to the Address table.")]
        public _Int32 BillToAddressID { get; private set; }

        [Required]
        [SqlNVarChar(50)]
        [DbColumn(Description = "Shipping method. Foreign key to ShipMethod.ShipMethodID.")]
        public _String ShipMethod { get; private set; }

        [SqlNVarChar(15)]
        [DbColumn(Description = "Approval code provided by the credit card company.")]
        public _String CreditCardApprovalCode { get; private set; }

        [Required]
        [SqlMoney]
        [DefaultValue(typeof(decimal), "0", Name = "DF_SalesOrderHeader_SubTotal")]
        [DbColumn(Description = "Sales subtotal. Computed as SUM(SalesOrderDetail.LineTotal)for the appropriate SalesOrderID.")]
        public _Decimal SubTotal { get; private set; }

        [Required]
        [SqlMoney]
        [DefaultValue(typeof(decimal), "0", Name = "DF_SalesOrderHeader_TaxAmt")]
        [DbColumn(Description = "Tax amount.")]
        public _Decimal TaxAmt { get; private set; }

        [Required]
        [SqlMoney]
        [DefaultValue(typeof(decimal), "0", Name = "DF_SalesOrderHeader_Freight")]
        [DbColumn(Description = "Shipping cost.")]
        public _Decimal Freight { get; private set; }

        [Required]
        [SqlMoney]
        [DbColumn(Description = "Total due from customer. Computed as Subtotal + TaxAmt + Freight.")]
        public _Decimal TotalDue { get; private set; }

        [SqlNVarCharMax]
        [DbColumn(Description = "Sales representative comments.")]
        public _String Comment { get; private set; }

        [_Computation]
        private void ComputeSalesOrderNumber()
        {
            SalesOrderNumber.ComputedAs((_String.Const("SO") + ((_String)SalesOrderID).AsSqlNVarChar(23)).IfNull(_String.Const("*** ERROR ***")));
        }

        [_Computation]
        private void ComputeTotalDue()
        {
            TotalDue.ComputedAs((SubTotal + TaxAmt + Freight).IfNull(_Decimal.Const(0)));
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_DueDate
        {
            get { return DueDate >= OrderDate; }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_Freight
        {
            get { return Freight >= _Decimal.Const(0); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_ShipDate
        {
            get { return ShipDate >= OrderDate | ShipDate.IsNull(); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_SubTotal
        {
            get { return SubTotal >= _Decimal.Const(0); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_TaxAmt
        {
            get { return TaxAmt >= _Decimal.Const(0); }
        }

        [_CheckConstraint]
        private _Boolean CK_SalesOrderHeader_Status
        {
            get { return IsValid(Status); }
        }

        private static _Boolean IsValid(_ByteEnum<SalesOrderStatus> status)
        {
            var byteExpr = (_Byte)status;
            return byteExpr >= _Byte.Const(1) & byteExpr <= _Byte.Const(6);
        }

        [_UniqueConstraint]
        private ColumnSort[] AK_SalesOrderHeader_SalesOrderNumber => new ColumnSort[] { SalesOrderNumber };

        [_DbIndex]
        private ColumnSort[] IX_SalesOrderHeader_CustomerID => new ColumnSort[] { CustomerID };
    }
}
